### ISSUE TITLE
module references aren't followed correctly

### ISSUE THREAD

**Issue by jdelic** [python/mypy#1741]
It seems that given an imported module and a reference to that module, mypy doesn't follow the reference.

Example:

``` python

# main.py
from test import mod

mod.blah()
for m in [mod]:
    m.blah()

# test/mod.py
def blah() -> None:
    print("blah")
```

`MYPYPATH=. bin/mypy main.py` will result in

```
main.py:5: error: "module" has no attribute "blah"
```

Note that mypy does not complain about `mod.blah()`, which is resolved correctly.

I apologize if this is a known bug, I read a lot of issues before writing this up, but wasn't able to find anything relating to this.

**Comment by gvanrossum** [comment]
IIUC the 'mod' object in main's global namespace is special, it knows exactly which module it references. But when you put a module in a list like your example (or do anything else with it other than plain assignment or an attribute reference), the resulting *type* becomes just the generic `module` and has no attributes except `__name__` and such.

What was your real-world use case?

**Comment by jdelic** [comment]
> What was your real-world use case?

I have a command-line application that, when started, calls `init_subsystem()` on multiple modules, each of which then does things like validate command-line parameters and loading plugins using `pkg_resources`. So my real code looks like this:

```python
from gopythongo import initializers, builders, versioners, assemblers, packers, stores

for s in [initializers, versioners, builders, assemblers, packers, stores]:
    s.init_subsystem()
```

I could just unroll the loop as a work-around... but is there a better solution?

```python
for s in [initializers.init_subsystem, builders.init_subsystem, ...]:
    s()
```

...will probably work, but strikes me as unwieldy.

**Comment by gvanrossum** [comment]
I see. Your special case defines an "interface" for each module to follow. Honestly I think the for-loop is a little clever here, so I would just unroll the loop. Or just add a `# type: ignore` to the call.

I know what mypy could do, in theory, but I don't think it's worth the effort of actually implementing it.

**Comment by fabianhjr** [comment]
Just found a case form an existing codebase,

The module `threading` has a function/attribute `enumerate` which is is used to check if there are other threads alive before exiting/restarting.

https://github.com/fabianhjr/home-assistant/blob/improvement-typing/homeassistant/__main__.py#L276,L281

**Comment by gvanrossum** [comment]
@fabianhjr What error do you get? I don't see how that code snippet relates to this issue yet -- the issue is about the type of `[mod]` being represented poorly (losing the fact that its values are all the `mod` module). I assume your problem is merely that the Python 3 stub for threading in typeshed doesn't have a definition for enumerate(). That's a typeshed bug which you can report here: https://github.com/python/typeshed (or better, submit a PR!).

**Comment by fabianhjr** [comment]
The error is:

```
homeassistant/__main__.py: note: In function "try_to_restart":
homeassistant/__main__.py:278: error: "module" has no attribute "enumerate"
```

**Comment by gvanrossum** [comment]
Yup, that has nothing to do with the issue here: even though the error message is superficially the same, the cause is unrelated.

**Comment by JukkaL** [comment]
More generally, module aliases aren't supported:
```
import module
alias = module
alias.foo  # not supported, even if module.foo exists
```

**Comment by ilevkivskyi** [comment]
The module aliasing now works, but iterating over list of modules seems like a very low priority. Also this is superseded by https://github.com/python/typeshed/issues/2147
